### PR TITLE
Use random generator from C++11 instead of from Boost

### PR DESCRIPTION
**Pull request by SunBlack**
There are still three occurences of Boost random generator: #2954 (metslib) #2955 (OutofcoreOctreeDiskContainer), #2988 (RangeLikelihood)

Solves partially #2803

### CODE REVIEW

File: common/include/pcl/common/impl/random.hpp [hunk 1]
```diff
@@ -40,14 +40,10 @@
...
-#include <pcl/pcl_macros.h>
-
 /////////////////////////////////////////////////////////////////////////////////////////////////////////
 template <typename T>
 pcl::common::UniformGenerator<T>::UniformGenerator(T min, T max, pcl::uint32_t seed)
   : distribution_ (min, max)
-  , generator_ (rng_, distribution_)
 {
   parameters_ = Parameters (min, max, seed);
   if(parameters_.seed != -1)
```
Comment (SunBlack) [inline common/include/pcl/common/impl/random.hpp]: Just saw a compiler warning related to this:
```
common/include/pcl/common/impl/random.hpp:49:23: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
   if(parameters_.seed != -1)
```
Even I didn't touched here anything: Should we fix this warning in this PR (and how?) or in a separate?

**Comment by SunBlack** [comment]
Tests seems to fail, because std random number generator produce other random numbers as boost random number generator with same seed. But not sure if this is always the reason - I even can't reproduce all test failures (Ubuntu just has 3, during Mac has 5). So it would be nice, if another can check this too.

**Comment by SergioRAgostinho** [comment]
Seems like you opened another can of worms :) This is a very big PR, so I'm in favor of splitting it up after we agree on good splitting points. After quickly skimming through the PR and getting acquainted with the new semantics behind `<random>` the use of `std::random_device` caught my eye.

One source [1] claims the method might actually throw on some implementations, source [2] only focuses on the aspect that it might still be pseudo random if there's no hardware support. I don't have experience with other platforms other than desktops so I'm not sure if the concern in [1] is applicable to us. I'm tempted to say that no. Opinions? 

Regarding the PR split, I feel the per module approach is worthy once more. I'm more interested in revising the core modules first and then progressing towards apps, examples and tools.

[1] - http://www.cplusplus.com/reference/random/random_device/
[2] - https://en.cppreference.com/w/cpp/numeric/random/random_device

**Comment by SunBlack** [comment]
> Seems like you opened another can of worms :)

Indeed, I wasn't expected this issues at the time I started :(


> This is a very big PR, so I'm in favor of splitting it up after we agree on good splitting points.

I was already thinking about it. My idea was to split everthing out, which causes tests to fail.

> One source [1] claims the method might actually throw on some implementations, source [2] only focuses on the aspect that it might still be pseudo random if there's no hardware support. I don't have experience with other platforms other than desktops so I'm not sure if the concern in [1] is applicable to us. I'm tempted to say that no. Opinions?

Well, I don't think [1] is an issue for us. I believe all modern platforms (mobile and desktop) can handle this. Maybe some specific platforms doesn't have it, but I'm nearly sure they are not part of the target group of the PCL.

> Regarding the PR split, I feel the per module approach is worthy once more. I'm more interested in revising the core modules first and then progressing towards apps, examples and tools.

Adjusting app, example and tools is easy, because they have no tests which depends on the result of generated values by the random generator, so they are straight forward during core components could be difficult to adjust.

Nevertheless we should discuss about another thing: Is it really necessary to generate random data within core classes? Usually you have separate factory classes in case you need random content. And btw: It is really ugly to have test which expect specific values of random generator.

**Comment by taketwo** [comment]
> It is really ugly to have test which expect specific values of random generator.

Fully agree. We can drop such tests, but then we may be left without any tests at all :D

**Comment by SergioRAgostinho** [comment]
> Adjusting app, example and tools is easy, because they have no tests which depends on the result of generated values by the random generator, so they are straight forward during core components could be difficult to adjust.

Let's go with the easy cases then first :+1: and progress to the more challenging problems as we progress.

> Is it really necessary to generate random data within core classes?

Likely not. But correcting this behaviour is not really mandatory at the moment.


> > It is really ugly to have test which expect specific values of random generator.
>
> Fully agree. We can drop such tests, but then we may be left without any tests at all :D

:sweat_smile:

File: common/include/pcl/common/impl/random.hpp (hunk 1 above)
Comment (taketwo) [reply common/include/pcl/common/impl/random.hpp]: We can fix it here. Just `static_cast<pcl::uint32_t>(-1)`.

Comment (SunBlack) [reply common/include/pcl/common/impl/random.hpp]: Just wondering about this:
https://github.com/PointCloudLibrary/pcl/blob/b4c785bf32b392b4570201e8fe9f97395bd98ba1/common/include/pcl/common/random.h#L163-L186

One constructor of `NormalGenerator` has as default seed `1` during the other one (via `Parameters`) has `-1` as default seed.

Comment (taketwo) [reply common/include/pcl/common/impl/random.hpp]: I think `-1` should be the default everywhere. However, thinking more about this, wouldn't it be better if `-1` meant "seed with random device" instead of "not seed at all".

Comment (SunBlack) [reply common/include/pcl/common/impl/random.hpp]: Moved discussion to #2985. I think it is better, if this behavior changes are not part of this PR.

File: simulation/src/range_likelihood.cpp [hunk 2]
```diff
@@ -1,5 +1,9 @@
-#include <GL/glew.h>
 #include <ctime>
+#include <random>
+
+#include <boost/math/distributions/normal.hpp>
```
Comment (SergioRAgostinho) [inline simulation/src/range_likelihood.cpp]: Is this one still required? I assumed the whitelist for now was only OutOfCoreDiskContainer and metslib. Is this one related?

File: surface/include/pcl/surface/mls.h [hunk 3]
```diff
@@ -732,15 +735,13 @@ namespace pcl
...
+      /** \brief Random number generator algorithm. */
+      mutable std::mt19937 rng_;
 
       /** \brief Random number generator using an uniform distribution of floats
         * \note Used only in the case of RANDOM_UNIFORM_DENSITY upsampling
         */
-      boost::shared_ptr<boost::variate_generator<boost::mt19937&,
-                                                 boost::uniform_real<float> >
-                       > rng_uniform_distribution_;
+      std::unique_ptr<std::uniform_real_distribution<>> rng_uniform_distribution_;
```
Comment (SergioRAgostinho) [inline surface/include/pcl/surface/mls.h]: Any idea why this a ptr? Is it because of the lazy/delayed initialization?

File: test/registration/test_correspondence_rejectors.cpp [hunk 4]
```diff
@@ -97,15 +97,13 @@ TEST (CorrespondenceRejectors, CorrespondenceRejectionPoly)
   pcl::transformPointCloud (cloud, target, t, q);
   
   // Noisify the target with a known seed and N(0, 0.005) using deterministic sampling
-  boost::mt19937 rng;
-  rng.seed (1e6);
-  boost::normal_distribution<> nd (0, 0.005);
-  boost::variate_generator<boost::mt19937&, boost::normal_distribution<> > var_nor (rng, nd);
+  std::mt19937 rng(1e6);
+  std::normal_distribution<> nd (0, 0.005);
```
Comment (SergioRAgostinho) [inline test/registration/test_correspondence_rejectors.cpp]: Better to just set the result type to `float` here and avoid the static cast below.

File: test/search/test_search.cpp [hunk 5]
```diff
@@ -543,15 +546,15 @@ void createIndices (std::vector<int>& indices, unsigned max_index)
 {
   // ~10% of the input cloud
   for (unsigned idx = 0; idx <= max_index; ++idx)
-    if (rand_uint () == 0)
+    if (rand_uint (rng) == 0)
       indices.push_back (idx);
    
-  boost::variate_generator< boost::mt19937, boost::uniform_int<> > rand_indices(boost::mt19937 (), boost::uniform_int<> (0, static_cast<int> (indices.size ()) - 1));
+  std::uniform_int_distribution<> rand_indices(0, static_cast<int> (indices.size ()) - 1);
```
Comment (SergioRAgostinho) [inline test/search/test_search.cpp]: I don't see the point behind this `static_cast<int>`. Just remove it.

File: test/surface/test_convex_hull.cpp [hunk 6]
```diff
@@ -321,14 +323,13 @@ TEST (PCL, ConvexHull_2dsquare)
   input_cloud->points.resize (input_cloud->width * input_cloud->height);
   
   //rng
-  boost::mt19937 rng_alg;
-  boost::uniform_01<boost::mt19937> rng (rng_alg);
-  rng.base ().seed (12345u);
+  std::mt19937 rng(12345u);
+  std::uniform_real_distribution<float> rd (0.0f, 1.0f);
```
Comment (SergioRAgostinho) [inline test/surface/test_convex_hull.cpp]: Set the constructor parameters appropriately here -1.0f to 1.0f, and then remove the math logic which is employed in lines 331 and 332.

File: test/surface/test_convex_hull.cpp [hunk 7]
```diff
@@ -380,16 +381,15 @@ TEST (PCL, ConvexHull_3dcube)
   input_cloud->height = 1;
   input_cloud->points.resize (input_cloud->width * input_cloud->height);
   
-  //rng
-  boost::mt19937 rng_alg;
-  boost::uniform_01<boost::mt19937> rng (rng_alg);
-  rng.base ().seed (12345u);
+  //rd
+  std::mt19937 gen(12345u);
+  std::uniform_real_distribution<float> rd (0.0f, 1.0f);;
```
Comment (SergioRAgostinho) [inline test/surface/test_convex_hull.cpp]: Same comment here.

File: tracking/include/pcl/tracking/boost.h [hunk 8]
```diff
@@ -1,46 +0,0 @@
...
- *
- */
-
-#pragma once
-
-#if defined __GNUC__
-#  pragma GCC system_header 
-#endif
-
-#include <boost/random.hpp>
```
Comment (SergioRAgostinho) [inline tracking/include/pcl/tracking/boost.h]: We might need to keep this file around for deprecation purposes, open for discussion. Since it is a header just emit pragma warning.

File: test/search/test_search.cpp (hunk 5 above)
Comment (SunBlack) [reply test/search/test_search.cpp]: `indices.size ()` is a `size_t`, so we have an unsigned type here. In case `size()` returns 0 we got with `static_cast<int> (indices.size ()) - 1` `-1` as value, during with without this cast we got `std::numeric_limits<size_t>::max()` => out of integer range => 0. Because this is test code it doesn't maybe matters. Just want to say we change behavior with it.

File: surface/include/pcl/surface/mls.h (hunk 3 above)
Comment (SunBlack) [reply surface/include/pcl/surface/mls.h]: It will be only initialized if `upsample_method_` is `RANDOM_UNIFORM_DENSITY`. And value range depends on `search_radius_`, so we have not a fixed value. So I could remove `unique_ptr`, but then we generate a object which we will discard without having used it.

File: simulation/src/range_likelihood.cpp (hunk 2 above)
Comment (SunBlack) [reply simulation/src/range_likelihood.cpp]: Missed this in whitelist `costFunction5` uses `pdf`/`cdf` which don't have a direct C++14 alias.

File: tracking/include/pcl/tracking/boost.h [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+ *
+ */
+
+#pragma once
+
+#if _MSC_VER
+  #pragma message ("This header is deprecated and will be removed in an upcoming release.")
+#elif __GNUC__
+  #warning ("This header is deprecated and will be removed in an upcoming release.")
+#endif
```
Comment (SergioRAgostinho) [inline tracking/include/pcl/tracking/boost.h]: Make use of this
https://github.com/PointCloudLibrary/pcl/blob/93f9e7233dfa87299212f6cbaa35dedfc1dd251b/common/include/pcl/pcl_macros.h#L227-L234

File: tracking/include/pcl/tracking/boost.h [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+#if defined __GNUC__
+#  pragma GCC system_header 
+#endif
```
Comment (SergioRAgostinho) [inline tracking/include/pcl/tracking/boost.h]: I would say we delete this block.

File: tracking/include/pcl/tracking/boost.h [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+#include <boost/random.hpp>
```
Comment (SergioRAgostinho) [inline tracking/include/pcl/tracking/boost.h]: Hmmm... I'm starting to notice a shortcoming on my request. I wanted to keep `pcl/tracking/boost.h` around, so it exists as a file, but I didn't want to include any boost headers inside it. If anyone was using this header for some reason it was just because of `boost/random.hpp`, then removing this line will still break their code, which means we need to keep it around. :/

**Comment by SergioRAgostinho** [comment]
The test are "legitimately" failing on windows. Shall we pull out/revert  the the failing case and associated class to a new PR so that we can address the unit test correction there?

**Comment by SunBlack** [comment]
I reverted changes to this test as I have no idea why they are failing. As we have already splitted other changes to other PRs from this PR because of failing tests, we could do this here too.

**Comment by SergioRAgostinho** [comment]
> As we have already splitted other changes to other PRs from this PR because of failing tests, we could do this here too.

Let's definitely do that. That way we can look at it from a finer scope.
